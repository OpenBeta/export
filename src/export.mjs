import fs from "node:fs";
import path from "node:path";
import { globby } from "globby";
import fm from "front-matter";
import { toString } from "mdast-util-to-string";
import { fromMarkdown } from "mdast-util-from-markdown";

export const convert_file = (baseDir, baseRegEx, fileName) => {
  const doc = fs.readFileSync(fileName, {
    encoding: "utf8",
  });

  const content = fm(doc);

  const mdast = fromMarkdown(content.body);

  const desc = [];
  for (const p of mdast.children.slice(1)) {
    if (p.type === "heading") break;
    const text = toString(p, { includeImageAlt: false });
    if (text.trim().length > 0) desc.push(text);
  }

  return {
    ...content.attributes,
    desc,
    src: path.posix.join(
      path.posix.basename(baseDir),
      fileName.replace(baseRegEx, "")
    ),
  };
};

/**
 * Export leaf areas (areas containing climbs)
 * @param baseDir
 */
export const to_jsonlines = async (contentDir, outputFile) => {
  // remove trailing slash
  const baseDir = contentDir.replace(/\/+$/g, "");

  // Get all leaf files, excluding dirs with only index.md
  const leafFiles = await globby([
    `${baseDir}/**/*.md`,
    `!${baseDir}/**/index.md`,
  ]);

  if (leafFiles.length === 0) {
    console.log("No files found");
    process.exit(0);
  }

  // Build a collection of leaf walls
  // 1. Remove file name from path
  // 2. Add to Set to remove duplicates
  const dirs = leafFiles.reduce((acc, curr) => {
    acc.add(path.dirname(curr));
    return acc;
  }, new Set());

  // de-dups
  const leafAreaPaths = [...dirs];

  // to make test deterministic
  leafAreaPaths.sort();

  fs.openSync(outputFile, "w");

  const baseRegEx = new RegExp(`^${baseDir}${path.sep}`, "g");
  leafAreaPaths.forEach((dir) => {
    const row = convert_file(baseDir, baseRegEx, dir + "/index.md");
    fs.appendFileSync(outputFile, JSON.stringify(row) + "\n");
  });

  console.log(`Records written: ${leafAreaPaths.length}`);
};
