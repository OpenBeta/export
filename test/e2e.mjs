import expect from "expect";
import { toMatchSnapshot } from "expect-mocha-snapshot";
import os from "os"
import fs from "fs";
import path from "path";

import { to_jsonlines } from "../src/export.mjs";

expect.extend({ toMatchSnapshot });

describe("End to end test", function () {
  it("Export test", async function () {
    const outDir = fs.mkdtempSync(path.posix.join("output", "test-"), {
      recursive: true,
    });
    const outfile = path.posix.join(outDir, "test.jsonlines");

    await to_jsonlines("examples", outfile);

    expect(fs.readFileSync(outfile).toString("utf-8")).toMatchSnapshot(this);
  });
});
