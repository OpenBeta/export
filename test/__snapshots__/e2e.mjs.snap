// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`End to end test Export test 1`] = `
"{\\"area_name\\":\\"Babaharimoss boulder\\",\\"metadata\\":{\\"area_id\\":\\"15fab2a5-c3ba-48c1-bb31-0bb2dc07c2d5\\",\\"mp_id\\":\\"109104861\\",\\"lng\\":-121.84467,\\"lat\\":45.49717},\\"desc\\":[],\\"src\\":\\"examples/Oregon/Mt. Hood/Lost Lake Boulders/Extra Crispy Boulders/Babaharimoss boulder/index.md\\"}
{\\"area_name\\":\\"Pine Line, The\\",\\"metadata\\":{\\"area_id\\":\\"410edef7-4896-4267-a645-d785d3d7855b\\",\\"mp_id\\":\\"113349046\\",\\"lng\\":-121.84464,\\"lat\\":45.49712},\\"desc\\":[],\\"src\\":\\"examples/Oregon/Mt. Hood/Lost Lake Boulders/Extra Crispy Boulders/Pine Line, The/index.md\\"}
{\\"area_name\\":\\"Quality Time Boulder\\",\\"metadata\\":{\\"area_id\\":\\"11234d3d-7c00-41e9-a3ce-e949500b29a9\\",\\"mp_id\\":\\"109252451\\",\\"lng\\":-121.84472,\\"lat\\":45.49716},\\"desc\\":[],\\"src\\":\\"examples/Oregon/Mt. Hood/Lost Lake Boulders/Extra Crispy Boulders/Quality Time Boulder/index.md\\"}
{\\"area_name\\":\\"Skid Mark Boulder\\",\\"metadata\\":{\\"area_id\\":\\"b6f23705-d30a-47e5-80b3-9bb9a9048e87\\",\\"mp_id\\":\\"109252443\\",\\"lng\\":-121.84467,\\"lat\\":45.49717},\\"desc\\":[],\\"src\\":\\"examples/Oregon/Mt. Hood/Lost Lake Boulders/Extra Crispy Boulders/Skid Mark Boulder/index.md\\"}
{\\"area_name\\":\\"Time Warp Boulder\\",\\"metadata\\":{\\"area_id\\":\\"05fc9c62-f6b1-4816-afc2-3ff15b12f606\\",\\"mp_id\\":\\"119055160\\",\\"lng\\":-121.84458,\\"lat\\":45.49759},\\"desc\\":[],\\"src\\":\\"examples/Oregon/Mt. Hood/Lost Lake Boulders/Extra Crispy Boulders/Time Warp Boulder/index.md\\"}
{\\"area_name\\":\\"White Rhino, The\\",\\"metadata\\":{\\"area_id\\":\\"50924451-49d4-444e-a16a-54665bd255de\\",\\"mp_id\\":\\"110624258\\",\\"lng\\":-121.84469,\\"lat\\":45.49741},\\"desc\\":[],\\"src\\":\\"examples/Oregon/Mt. Hood/Lost Lake Boulders/Extra Crispy Boulders/White Rhino, The/index.md\\"}
{\\"area_name\\":\\"kids in the hall boulder\\",\\"metadata\\":{\\"area_id\\":\\"36c29bab-b478-407d-a9b3-95e3be10e774\\",\\"mp_id\\":\\"109104867\\",\\"lng\\":-121.84467,\\"lat\\":45.49717},\\"desc\\":[],\\"src\\":\\"examples/Oregon/Mt. Hood/Lost Lake Boulders/Extra Crispy Boulders/kids in the hall boulder/index.md\\"}
{\\"area_name\\":\\"Quirky Combat Wall\\",\\"metadata\\":{\\"lng\\":-121.568433,\\"lat\\":45.405378,\\"area_id\\":\\"200bb0e7-62dd-43c3-b1de-3110e1b7f675\\",\\"mp_id\\":\\"\\"},\\"desc\\":[\\"Keep going left of K9 Shangai for about 150 yards. The faint trail follows the edge of the wall until you get to the next group of scrubbed routes.\\",\\"The trail is on slopping very soft dirt and it does not take much to trigger a slide. Thread lightly.\\",\\"All these routes are fairly new and even though great care was take to remove obvious loose rocks, some more may still be present. Helmets are strongly recommended. Some of these new routes might be in need of a light re-scrubbing. You should find brushes in the tree-mounted box along the trail. You are welcome to take one with you during your visit.\\"],\\"src\\":\\"examples/Oregon/Mt. Hood/Petes Pile/Quirky Combat Wall/index.md\\"}
"
`;
